File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.cpp
class Solution {
public:
    int minDeletions(string s) {
        //Approach_02: using priority_queue:
        vector<int>freq(26,0);
        for(char c:s){
            freq[c-'a']++;
        }
        priority_queue<int> max_heap;
        for (int i = 0; i<26; i++){
            if (freq[i] > 0) {
                max_heap.push(freq[i]);
            }
        }
        int cnt = 0;
        while(max_heap.size()>1)
        {
            int top1 = max_heap.top();
            max_heap.pop();
            if(top1 == max_heap.top()){
                if(top1-1 > 0){
                    max_heap.push(top1-1);
                }
                cnt++;
            }
        }
        return cnt;
    }
};

File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.java
class Solution {
    public int minDeletions(String s) {
        // Time:O(n+k^2logk) and Space:O(k)
        int freq[] = new int[26];
        for(int i = 0;i<s.length();i++){
            freq[s.charAt(i)-'a']++;
        }
        PriorityQueue<Integer> max_heap = new PriorityQueue<>(Collections.reverseOrder());
        for(int i=0;i<26;i++){
            if(freq[i]>0){
                max_heap.add(freq[i]);
            }
        }
        int cnt = 0;
        while(max_heap.size()>1)
        {
            int topElement = max_heap.remove();
            if(topElement == max_heap.peek()){
                if(topElement - 1 > 0){
                    max_heap.add(topElement-1);
                }
                cnt++;
            }
        }
        return cnt;
    }
}